src/storage/v3/storage.cpp

```diff
@@ -515,38 +491,46 @@ ResultSchema<VertexAccessor> Storage::Accessor::CreateVertexAndValidate(
     return {std::move(*maybe_schema_violation)};
   }
   OOMExceptionEnabler oom_exception;
-  auto gid = storage_->vertex_id_.fetch_add(1, std::memory_order_acq_rel);
+  // Extract key properties
+  std::vector<PropertyValue> primary_properties;
+  for (const auto &[property_id, property_value] : properties) {
+    if (storage_->schemas_.IsPropertyKey(primary_label, property_id)) {
+      primary_properties.push_back(property_value);
+    }
+  }
+
+  // Get secondary properties
+  std::vector<std::pair<PropertyId, PropertyValue>> secondary_properties;
+  for (const auto &[property_id, property_value] : properties) {
+    if (!storage_->schemas_.IsPropertyKey(primary_label, property_id)) {
+      secondary_properties.emplace_back(property_id, property_value);
+    }
+  }
+
   auto acc = storage_->vertices_.access();
   auto *delta = CreateDeleteObjectDelta(&transaction_);
-  auto [it, inserted] = acc.insert({Vertex{Gid::FromUint(gid), delta, primary_label}});
+  auto [it, inserted] = acc.insert({Vertex{delta, primary_label, primary_properties, labels, secondary_properties}});
```

Comment:
Why not to move the `primary_properties` and `secondary_properties` into the vertex? They are already copied to the stack, let's let the `Vertex` steal the resources.

Usually this wouldn't matter, however I think `CreateVertex` can be on a hot path, especially for loading data into Memgraph.